### PR TITLE
[MXNET-984] Java NDArray Documentation Generation

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
Since the new Operator API is formed. This Documentation Generation changes slightly on the NDArrayMarcos and shows the structure of the expected outcome.
@nswamy @yzhliu @andrewfayres

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by andrewfayres** [comment]
Do you have an example of the output docs that you can share?

**Comment by lanking520** [comment]
> Do you have an example of the output docs that you can share?

Yes! https://gist.github.com/lanking520/f9ed0eeaa45f2364471f1f3c6c9d598f

**Comment by andrewfayres** [comment]
This is very similar code to the other APIDocGenerator class one level up. It looks like all that really changes are package names and stuff. Is there anything that prevents us from writing a base class and two subclasses instead of having a lot of near identical code?

**Comment by Roshrini** [comment]
@mxnet-label-bot [Java, Doc, pr-awaiting-review]

File: scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/APIDocGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    val absFuncs = absClassFunctions.filterNot(_.name.startsWith("_"))
+      .filterNot(ele => notGenerated.contains(ele.name))
+      .map(absClassFunction => {
+      val scalaDoc = generateAPIDocFromBackend(absClassFunction)
+      val defBody = generateAPISignature(absClassFunction)
+      s"$scalaDoc\n$defBody"
+    })
+    val packageName = "NDArrayBase"
+    val apacheLicence = "/*\n* Licensed to the Apache Software Foundation (ASF) under one or more\n* contributor license agreements.  See the NOTICE file distributed with\n* this work for additional information regarding copyright ownership.\n* The ASF licenses this file to You under the Apache License, Version 2.0\n* (the \"License\"); you may not use this file except in compliance with\n* the License.  You may obtain a copy of the License at\n*\n*    http://www.apache.org/licenses/LICENSE-2.0\n*\n* Unless required by applicable law or agreed to in writing, software\n* distributed under the License is distributed on an \"AS IS\" BASIS,\n* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n* See the License for the specific language governing permissions and\n* limitations under the License.\n*/\n"
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/APIDocGenerator.scala]: please look at mdespriee's he has fixed this blob of text into lines.

File: scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/APIDocGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    val scalaStyle = "// scalastyle:off"
+    val packageDef = "package org.apache.mxnet.javaapi"
+    val imports = "import org.apache.mxnet.annotation.Experimental"
+    val absClassDef = s"abstract class $packageName"
+    val finalStr = s"$apacheLicence\n$scalaStyle\n$packageDef\n$imports\n$absClassDef {\n${absFuncs.mkString("\n")}\n}"
+    val pw = new PrintWriter(new File(FILE_PATH + s"$packageName.scala"))
```
Comment (piyushghai) [inline scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/APIDocGenerator.scala]: Should the extension be .scala ? Or .java ?

**Comment by andrewfayres** [comment]
I don't have a diff of the code in front of me but it looks like it's identical except for a few things like the name of the package etc. Maybe add a gist of the differences to get some ideas but there's got to be a better way to do this than replicating a whole class.

Maybe this is more difficult than I think it is but it looks like you could take the existing generator and replace the hardcoded org.apache.mxnet with a class variable. When you want to run the java doc generation we'd set that variable to org.apache.mxnet.javaapi.

**Comment by lanking520** [comment]
@andrewfayres thanks for your feedback on my concern. Then I would try to make some changes on the Scala APIGenerator. I will try to reuse the code as much as possible but please expected something big change on the existing code base.

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 3]
```diff
@@ -157,19 +203,48 @@ private[mxnet] object APIDocGenerator{
     s"$experimentalTag\ndef ${func.name} (${argDef.mkString(", ")}) : $returnType"
   }
 
+  def generateJavaAPISignature(func : absClassFunction) : String = {
+    // scalastyle:off
```
Comment (andrewfayres) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: Why are we turning this off here?

**Comment by lanking520** [comment]
@andrewfayres Thanks for your comments, all documentation generation looks fine. Please see the following gist just in case https://gist.github.com/lanking520/501eebcdf27b0a0bcb5444eee37baa75

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 4]
```diff
@@ -54,8 +54,46 @@ private[mxnet] object APIDocGenerator{
     org.apache.commons.codec.binary.Base64.encodeBase64URLSafeString(digest)
   }
 
-  def absClassGen(FILE_PATH : String, isSymbol : Boolean) : String = {
-    // scalastyle:off
+  def fileGen(filePath : String, packageName : String, packageDef : String,
+              absFuncs : List[String]) : String = {
+    // Copied from @mdespriee in PR #12489
+    val apacheLicence =
```
Comment (piyushghai) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: nit : typo : ```apacheLicense```

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 5]
```diff
@@ -52,8 +54,46 @@ private[mxnet] object APIDocGenerator{
     org.apache.commons.codec.binary.Base64.encodeBase64URLSafeString(digest)
   }
 
-  def absClassGen(FILE_PATH : String, isSymbol : Boolean) : String = {
-    // scalastyle:off
+  def fileGen(filePath : String, packageName : String, packageDef : String,
+              absFuncs : List[String]) : String = {
+    // Copied from @mdespriee in PR #12489
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: you can drop this

**Pull request merged**